patterns/1-initial/modular-code.md

```diff
@@ -8,54 +8,69 @@ The lack of modularization in the software architecture prevents reuseability, a
-* Acknowledge and drive acceptance for the cost of modularization. Build time into the release schedule for this.
-* Companies moving to use more open source code will appreciate modularity more over time.
-* Mitigate risk and fear of quality degradation from accepting InnerSource contributions.
-* Establish a checklist of elements to be checked off to classify a component as reusable.
-    - There could be requirements on tests, tools and documentation before considering a component as reusable
-    - Introduce refactoring as organic effort and strive for it to become part of the culture. Refactoring leads to more modular code and reduction of technical debt.
-* Establish standards on testing methodology, labeling of repos.
+Modular code is far more reusable. Decoupling functionality into smaller narrow-focused modules eases
+and therefore promotes external contributions and onboarding of newcomers.
+
+* Offer education.
+  * On why: Lack of modularity is technical debt. It increases costs but hides them by shifting these into the future.
+  * On how: Modular code is well-understood; there is a lot of literature in favor of this.
+  * Select certain "success projects", teams that will develop reusable code and demonstrate the long term success. This can help motivate others (they see what is possible and what is in it for them). Transparency is critical.
+* Introduce refactoring as organic effort and strive for it to become part of the culture.
+  * Acknowledge and drive acceptance for the investion of modularization. Build time into the release schedule for this.
+* Search for modularizable functionalities:
+  * Identify common functionalities and code blocks with potential for profitable modularity.
+  * Establish a checklist of elements to be checked off to classify a component as reusable.
+    There could be requirements on tests, tools and documentation before considering a component as reusable.
+    Establish standards on testing methodology, labeling of repos.
+* Eat the elephant in small bites. Break modularization down into baby steps, like
+  1. find internal would-be customers. Better ask them over guessing on your own.
+  1. define decoupling scope and prefferred decoupling methods,
+  1. develop automated testing protection
+  1. document usage of modules.
+  1. introduce semantic versioning
+  1. use a dependency management software such as pip, nuget, npm
+* More information and techniques are available from the book **Working Effectively with Legacy Code** by Michael Feathers and **Refactoring: Improving the Design of Existing Code** by Martin Fowler
 
 ## Resulting Context
 
-**Pattern Idea**: Time is spent making the shared code modular so it can be reused.  
+* Modular code facilitates division of paralell labour (as required for work in larger teams, or as an open source project). Modular projects enable team members to focus on their own smaller assigned tasks.
+* Modular code makes Agile implementation easier.
```

Comment:
What do you mean by this? Sounds similar to the "smaller assigned tasks" above.

Reply:
Smaller grains make sand more fluid.

Agile relies on small iterations. Smaller codebase bits and atomic unit tests provide more alternative combinations which in turn allows to pack more work into timeboxed sprints. It also helps when replanning. And smaller bits have smaller scopes (less to learn and document), so user stories get more specific (less to discuss).

For the same functionality, you'll need to change more bits, but the modularity helps to materialize sooner (earlier and more reliable validations) and to prioritize the most relevant bits (separate value from waste).

Reply:
It's all about smaller tasks, but the value provided by making tasks smaller is different.

Reply:
> Smaller grains make sand more fluid.

⬆️ love that! Had not heard that expression before but it is a great mental picture.

I can see that it might be hard to pack your whole statement into row 72. Still wondering if we could add some of it? e.g.

> Modular code makes Agile implementation easier e.g. by allowing for smaller iterations and through that earlier customer feedback.

Reply:
Saw the changes the Agile bit. Fantastic!